### DISCUSSION TITLE
Extending functionalities of OrchardCore

### DISCUSSION THREAD

**Discussion by rwxzig** [OrchardCMS/OrchardCore#14325] [Q&A] [answered]
Hello everyone, 
I have been reading and trying to implement various scenarios in Orchard so that I can be more familiar with the framework. 

Suppose that we have a school staff management system. We have teachers, principal and classes. Principal can only create classes that start within a time span. When a teacher is assigned to that class, the teacher gets a notification (vial email or something else) where the teacher can approve the class or reject it, in both cases the principal receives a notification.

1. How would I implement a safeguard that stops for instance principal from creating classes that are out of the given timespan?
2. How would I implement the approve and reject feature for a given content item? (In this way I can learn to extend the functionalities beyond what is given out of the box).
3. How would it be possible to check if the teacher is not already signed up for another class in that particular time.
4. Is it possible add or modify a content item from a page, without needing to do it from admin panel. As an example, If I have a calendar and I want to add items from that view. 

I created this scenario just so that I can learn on how to extend the functionalities of the framework and they are purely fictional, so even if it does not make sense, this is just for educational purpose. 

From what I have read, **workflows** can help in this use case, but is there any other preferred way or any other way in general.
I am not looking for code implementation, but a brief explanation would be greatly appreciated.

Also a nice feature would be to come out of the box, commenting like in Drupal with their Comments Module.

*Edit 1:*
I see that there is a discussion regarding comments on [this](https://github.com/OrchardCMS/OrchardCore/discussions/7552) thread and issue [#2942](https://github.com/OrchardCMS/OrchardCore/issues/2942)

*Edit 2:*
Added question 4.

*Edit 3:*
Given that OrchardCore can be a headless CMS, it must be possible and easy to implement nr 4, I just need to dig more. 😄 

I apologize in advance if this is the wrong place to ask these questions.

Best regards.

**Comment by hishamco** [comment]
> Also a nice feature would be to come out of the box, commenting like in Drupal with their Comments Module.

I remembered Comments was in Orchard CMS previously, it's good to bring it back, especially in `TheBlogTheme`

**Comment by Piedone** [comment, marked as answer]
1/3. The basic answers to your questions 1 and 3 are "do a SQL query of the corresponding content items/entities to check". Since we're talking about a very specific use-case that Orchard doesn't support out of the box (since it's not a school management system) the approach here would be to model your data, and then query and interact with it. See [the tutorials](https://docs.orchardcore.net/en/latest/docs/resources/tutorials/) for the basics of data management and content CRUD that are relevant here.
2. There's more than one way, but I'd perhaps create a content part to manage this (assuming you store classes in content items), with a property storing the approval state, and permissions to let only the principal approve or deny. You can send e-mails with `ISmtpService`. Workflows can be utilized but I'd rather write some code for this; or you can combine and create a Workflows Event that raises an approved/rejected event, and then you can tie e.g. e-mail sending on it (less explicitly would be the Content Published event if you also publish the item on approve/reject).
4. Yes, but you either need to display the editor on the frontend and handle its POST, or rather use a custom frontend editor to being with. The [Training Demo Module](https://github.com/Lombiq/Orchard-Training-Demo-Module) has sample code for most of this.
